Addition of  partition_ddf to Ktk

usha-nemani-by opened the pull request:
This PR is continuation to PR#357 , to add more changes based on the discussion.

Addition of  partition_ddf func to Ktk

- [ ] Addition of  partition_ddf  func to Ktk
- [ ] New test cases addition. 
- [ ] Changelog entry

usha-nemani-by commented:
@fjetter Comment :

I'm still not entirely sold on this API. The repartition ratio itself was more or less a workaround I introduced back in the day to justify the repartitioning in update_dataset_from_ddf. Taking a step back, we'd be introducing two functions here which would always be called together

```
ratio_graph = calculate_repartition_ratio(..., target_nrows_per_partition, frac)
ratio = ratio_graph.compute()
new_ddf = repartition_ddf(ddf, repartition_ratio=ratio)

```
I'm wondering if we should not instead introduce one function

```
def repartition_ddf(ddf, partition_size: int):
    # We would block here
    npartitions = _calc_npartitions(...).compute()
    return ddf.repartition(npartitions=npartitions)
```

The implementation would be more or less the same but it feels to me as if the API was more user friendly. The only reason why two separate calls would be justified if we'd wanted to stick to the lazy execution.

lr4d commented:
I had already approved the (previous) PR. @fjetter Can you share your thoughts?

fjetter commented:
My thoughts are effectively written above. the repartition ratio is not a good api I believe, since it's rather obscure without delivering actual value to the user in terms of functionality.
I introduced the parameter back in the day on application level and figured that repartitioning is essential for shuffling but I didn't have a proper tool to describe it better and just moved the application level config to the library. However, the repartition ratio has little value compared to the plain `npartitions` of task. The value we want to offer in this PR is that we may calculate rapidly how many partitions are requested by the user. In effect we're offering a faster alternative to the existing api of dask but dask doesn't have a repartition ratio. 
Therefore, I argue that that we should rather use quantities already familiar. 
After all, the task/story we are working on at the moment is about this exact problem. Repartiton ratio is a fragile abstraction. Number of rows (or size in MB) is much more tangible and robust

lr4d commented:
>  Therefore, I argue that that we should rather use quantities already familiar.

I agree

marco-neumann-by commented on kartothek/io/dask/dataframe.py:
```diff
@@ -606,3 +604,67 @@ def hash_dataset(
+    Calculate a `repartition_ratio` according to the dataset input and `target_nrows_per_partition` (the desired
+    number of rows of each dataframe partition).
+
+
```
1. please add a note that this function performs IO.
2. "calculate" could be read by many users as "compute on client". I would prefer something like `compute`.

usha-nemani-by replied:
`calculate_repartion_ratio()` function is removed.

marco-neumann-by commented on kartothek/io/dask/dataframe.py:
```diff
@@ -606,3 +604,67 @@ def hash_dataset(
+    ddf
+        dask.Dataframe
+    repartition_ratio
+        calculated repartition ratio
```
Please linke `calculate_repartition_ratio`

usha-nemani-by replied:
We are removing this functionality. Please review the latest code changes.

marco-neumann-by commented on kartothek/io/dask/dataframe.py:
```diff
@@ -606,3 +604,67 @@ def hash_dataset(
+    ddf: dd.DataFrame, repartition_ratio: SupportsFloat
+) -> dd.DataFrame:
+    """
+    Repartition a dask DataFrame according, dividing the current number of partitions by `repartition_ratio`.
```
What exactly is a "repartition ratio"? Neither this nor the other function explain this term.

marco-neumann-by replied:
I should have read the PR comments before, sorry. I see you already more or less agreed that this term is somewhat fragile / non-user-friendly.

usha-nemani-by replied:
The logic is updated completely. Please review new logic and give suggestions.

marco-neumann-by commented on tests/io/dask/dataframe/test_repartition.py:
```diff
@@ -0,0 +1,106 @@
+    ).all()
+
+
+@pytest.mark.parametrize("repartition_ratio", (3.88, 0.25, 7, 1.0))
```
where do these numbers come from?

usha-nemani-by replied:
They are random numbers - >1,  <1 , ==1

usha-nemani-by commented:
>Therefore, I argue that that we should rather use quantities already familiar.
> After all, the task/story we are working on at the moment is about this exact problem. Repartiton ratio is a fragile 
abstraction. Number of rows (or size in MB) is much more tangible and robust


Based on above discussion, I modified the functionality.
I have removed the `calculate_repartition_ratio()` function.
Modified the `repartition_ddf()` functionality  as `compute_repartition_ddf()`to compute  the repartition_ratio and compute `target_npartitions` and finally repartition the ddf.
I felt that `compute_repartition_ddf()` itself is a small function with 4 steps:
1. Collect dataset_metadata
2. compute repartition_ratio using `target_nrows_per_partition`
3. calculate `target_npartitions` using above ratio.
4. repartition the ddf and optimize it.


If I divide the function into 2 smaller funcs again, I need to send all the parameters from `compute_repartition_ddf()` to the `_calc_target_npartitions()` as well.

```
def compute_repartition_ddf(ddf, dataset_factory, target_nrows_per_partition, frac, predicates , ):
      code block here
     _calc_target_npartitions(ddf, dataset_factory, target_nrows_per_partition, frac, predicates)
  
```

I have updated the description for the same.
Please confirm if the single function looks fine.

lr4d commented on tests/io/dask/dataframe/test_repartition.py:
```diff
@@ -0,0 +1,98 @@
+        (pd.DataFrame({"A": range(10), "part": range(10)}), 1),  # rep_ratio = 1.0
+    ),
+)
+def test_calculate_repartition_ratio_approximate_match(
```
```suggestion
def test_repartition_ddf_approximate_match(
```

lr4d commented on kartothek/io/dask/dataframe.py:
```diff
@@ -606,3 +607,53 @@ def hash_dataset(
+        target num of rows per partition
+    dataset_metadata_frac
+        This fraction is helps in collecting metadata on a fraction(sample) of the ddf, instead of the whole ddf.
+
+
```
I was thinking if we should add `chunksize`as dask allows i.e. the uncompressed size per partition. That is probably more valuable than `target_nrows_per_partition`

usha-nemani-by replied:
We should fix about this Parameter as this would be the customer facing one. 
@fjetter , any thoughts here?

lr4d commented on kartothek/io/dask/dataframe.py:
```diff
@@ -606,3 +607,53 @@ def hash_dataset(
+        dask.DataFrame
+    dataset_factory
+        dataset factory
+    predicates
+        Optional , but predicates can be applied while collecting_dataset_metadata
```
```suggestion
    predicates
        Optional , but predicates can be applied while collecting_dataset_metadata.
        Note that predicates will only be applied during query planning, predicate pushdown is not (yet) supported.
        Thus, this is only useful if providing predicates involving indexed columns.
```

lr4d commented on tests/io/dask/dataframe/test_repartition.py:
```diff
@@ -0,0 +1,98 @@
+    ).all()
+
+
+#
+# @pytest.mark.parametrize("repartition_ratio", (3.88, 0.25, 7, 1.0))
+# def test_optimize_with_diff_repartition_ratios(repartition_ratio, store_factory):
+#     dfs = [pd.DataFrame({"A": range(100), "part": ["A", "B", "C", "D"] * 25})]
+#     store_dataframes_as_dataset__iter(
+#         dfs, store=store_factory, dataset_uuid="test", partition_on=["part"],
+#     )
+#     ddf = read_dataset_as_ddf(store=store_factory, dataset_uuid="test")
+#     optimized_ddf = dask.optimize(ddf, optimizations=[dask.dataframe.optimize])[0]
+#     repartitioned_ddf = repartition_ddf(ddf, repartition_ratio=repartition_ratio)
+#     if repartition_ratio >= 1:
+#         assert len(repartitioned_ddf.dask) <= len(
+#             optimized_ddf.dask
+#         )  # assert the number of tasks in dask graph
+#     else:
+#         assert len(repartitioned_ddf.dask) > len(optimized_ddf.dask)
```
We'll want to replace this with something analogous for repartition_ddf (target_nrows or chunksize)

usha-nemani-by replied:
The first two testcase parameters are ddf and  target_nrows . So I was thinking, is this testcase required at all?

lr4d commented:
> If I divide the function into 2 smaller funcs again, I need to send all the parameters from repartition_ddf() to the _calc_target_npartitions() as well.

I think it makes sense to isolate the "calculate" logic from the repartitioning logic though.

usha-nemani-by commented:
> > If I divide the function into 2 smaller funcs again, I need to send all the parameters from repartition_ddf() to the _calc_target_npartitions() as well.
> 
> I think it makes sense to isolate the "calculate" logic from the repartitioning logic though.

I have isolated the calculation logic. Please verify and confirm.

usha-nemani-by commented:
@fjetter / @lr4d , 
In the previous PR  (#357 ), there was a suggestion to call   `repartition_ddf()` in   [https://github.com/JDASoftwareGroup/kartothek/blob/9e4909102c0b5d95b36d25136a6479fb71c644ab/kartothek/io/dask/dataframe.py#L336](url)  

If that is the case, do I  need to  remove the `repartition_ratio` from input params, and change the signature of that function?

[https://github.com/JDASoftwareGroup/kartothek/blob/9e4909102c0b5d95b36d25136a6479fb71c644ab/kartothek/io/dask/dataframe.py#L431](url)

lr4d commented:
After discussion involving Usha, Florian and I, it was decided to close this until we agree on an API interface for this functionality
